### PR TITLE
Implement data flow map reduce tree construction

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
* Bulk of the logic can be found in `DataFlowTreeFactory`
* A couple of classes were renamed where the only changes are some additional documentation.

## Testing
* Ran locally on PRISM study with the following request:
```
{
    "filters": [
        {
            "entityId": "EUPATH_0000609",
            "type": "numberRange",
            "variableId": "EUPATH_0000047",
            "min": 3.2,
            "max": 15.4
        },
        {
            "entityId": "EUPATH_0000776",
            "type": "numberRange",
            "variableId": "EUPATH_0000135",
            "min": 0,
            "max": 500.5
        }
    ],
    "outputVariableIds": [
        "OBI_0001169",
        "EUPATH_0000162",
        "EUPATH_0000317"
    ]
}
```

### CODE REVIEW

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java [hunk 1]
```diff
@@ -127,6 +133,38 @@ public static void produceTabularSubset(DataSource dataSource, String appDbSchem
     }, FETCH_SIZE_FOR_TABULAR_QUERIES);
   }
 
+  /**
+   * Writes to the passed output stream a "tabular" result.  Exact format depends on the passed
+   * responseType (JSON string[][] vs true tabular). Each row is a record containing
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java:139]: maybe the passed in formatter controls the json v. tabular output?

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java [hunk 2]
```diff
@@ -127,6 +133,38 @@ public static void produceTabularSubset(DataSource dataSource, String appDbSchem
...
+   * the primary key columns and requested variables of the specified entity.
+   *
+   * @param study           study context
+   * @param outputEntity    entity type to return
+   * @param outputVariables variables requested
+   * @param filters         filters to apply to create a subset of records
+   */
+  public static void produceTabularSubsetFromFile(Study study, Entity outputEntity,
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java:147]: maybe the method name should be `...FromFiles`.  as it is, i am looking for an input file

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java [hunk 3]
```diff
@@ -127,6 +133,38 @@ public static void produceTabularSubset(DataSource dataSource, String appDbSchem
     }, FETCH_SIZE_FOR_TABULAR_QUERIES);
   }
 
+  /**
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java:137]: would be nice if comment indicates that this is the top level of a recursive map-reduce blah blah blah

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                                      Entity outputEntity,
+                                      List<Filter> filters,
+                                      List<Variable> outputVariables,
+                                      Study study) {
+    if (!outputVariables.stream().allMatch(var -> var.getEntity().equals(outputEntity))) {
+      throw new IllegalArgumentException("All output variables must be associated with output entity.");
+    }
+    try {
+      LOG.info("Creating map-reduce data-flow tree rooted with output entity: " + outputEntity.getId());
+      final TreeNode<Entity> parent = prunedEntityTree.findFirst(entity -> entity.equals(outputEntity));
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java]: i dont know i could state what is going on in this method, so i'd suggest inline comments.

i was wondering why we have to do a find to locate our node in the tree.  could somebody at a higher level be sure we always get the relevant subtree?  just a thought.

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+   * Recursively depth-first traverse each child and then the parent.
+   *
+   * @param originalRoot The original entity tree's root. This is needed to find the parent of our current node.
+   * @param currentTraversalNode Node that is currently being traversed.
+   * @param previousNode Node that was previously traversed. We keep track of this to avoid traversing from whence we came.
+   * @param filters List of all filters in request, used in construction of the {@link SubsettingJoinNode} corresponding to this node.
+   * @param study Study that subsetting data is requested for.
+   * @return
+   */
+  private SubsettingJoinNode rerootTree(TreeNode<Entity> originalRoot,
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java]: i am having trouble understanding this method too.

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/FormattedTabularRecordStreamer.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+   *                         variables.
+   * @param idIndexStream    Stream of ID indexes, indicating which entity records to output.
+   * @param ancestorStream   Stream of ancestors, used to merge in ancestors of target output entities in output records.
+   * @param entity           Entity type of records which will be output.
+   */
+  public FormattedTabularRecordStreamer(List<Iterator<VariableValueIdPair<?>>> valuePairStreams,
+                                        Iterator<Long> idIndexStream,
+                                        Iterator<VariableValueIdPair<List<Long>>> ancestorStream,
+                                        Entity entity) {
+    this.valuePairStreams = valuePairStreams.stream()
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/reducer/FormattedTabularRecordStreamer.java:39]: its a little confusing that the instance var and param have same name but are diff types.  maybe change the name of one, and explain why u do this transform

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/FormattedTabularRecordStreamer.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    List<String> record = new ArrayList<>();
+    // Only add ancestors if entity has ancestors associated with it.
+    if (!entity.getAncestorEntities().isEmpty()) {
+      if (!ancestorStream.hasNext()) {
+        throw new NoSuchElementException("No tabular records remain in stream.");
+      }
+      while (ancestorStream.peek().getIdIndex() < currentIdIndex) {
+        ancestorStream.next();
+      }
+      // Add all ancestor IDs
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/reducer/FormattedTabularRecordStreamer.java:71]: need to check the order of id and ancestors.   is it:  id, parent, grandparent, greatgp   or greatgp, gp, parent, id?

the binary file dumper seems to put the id at the left

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java (hunk 4 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java]: We could. I'd just need to update the FgpUtil `TreeNode` to store a reference to its parent.

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java (hunk 5 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java]: Some direction might be helpful here or perhaps a conversation

**Review by steve-fischer-200 — APPROVED** [review]

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java [hunk 8]
```diff
@@ -52,6 +58,7 @@
 public class FilteredResultFactory {
 
   private static final Logger LOG = LogManager.getLogger(FilteredResultFactory.class);
+  private static final DataFlowTreeFactory DATA_FLOW_TREE_FACTORY = new DataFlowTreeFactory();
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java:61]: Wondering if this needs to be static?  Just means we have to be careful about thread safety, but empty constructor means we can cheaply construct when needed.

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+public class DataFlowTreeFactory {
+  private static final Logger LOG = LogManager.getLogger(DataFlowTreeFactory.class);
+
+  public TreeNode<DataFlowNodeContents> create(TreeNode<Entity> prunedEntityTree,
+                                               Entity outputEntity,
+                                               List<Filter> filters,
+                                               Study study) {
+    final TreeNode<Entity> outputNode = prunedEntityTree.findFirst(entity -> entity.equals(outputEntity));
+    // In lieu of a pointer up the tree, we use this function to traverse from the original root to a node's parent.
+    final Function<TreeNode<Entity>, TreeNode<Entity>> parentMapper = child ->
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java:36]: I wonder if you want to return an Optional here vs null for the root entity.  Could enable later logic like parent.ifPresent(parent -> /*do something*/) + NPE prevention.

**Review by ryanrdoherty — APPROVED** [review]

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java (hunk 9 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java:36]: Makes sense, addressed this in coming PR.

**Pull request merged**